alf/algorithms/algorithm.py

```diff
@@ -1275,11 +1281,65 @@ def _train_experience(self,
+            #
+            # Suppose we gather_all() gives us env_ids = [0, 1] and positions =
+            # [7, 7] for batch_info. This means that before the exprience is
+            # chopped by mini_batch_length (reshape), there are two trajectories
+            # for env 0 and env 1. Assuming there are 12 steps in each of the
+            # trajectories, and the mini_batch_length is 3.
+            #
+            # After the choppping (reshape), it is expected to have 8
+            # trajectories in experience (because each original trajectory is
+            # now chopped into 4 new trajectories of length 3), and therefore we
+            # would like to adjust the batch_info to correctly reflect that. The
+            # result would be
+            #
+            # env_ids =   [0,  0,  0,  0,  1,  1,  1,  1]
+            # positions = [7, 10, 13, 16,  7, 10, 13, 16] (before circular())
+            num_envs = batch_info.env_ids.shape[0]
+            replay_buffer = batch_info.replay_buffer
+            num_mini_batches_per_original_traj = length // mini_batch_length
+            batch_info = BatchInfo(
+                env_ids=batch_info.env_ids.repeat_interleave(
+                    num_mini_batches_per_original_traj),
+                positions=replay_buffer.circular(
+                    batch_info.positions.repeat_interleave(
+                        num_mini_batches_per_original_traj) + torch.arange(
+                            0, length, mini_batch_length).repeat(num_envs)),
+                replay_buffer=replay_buffer)
+
+            # Treatment 2: Adjust the mini_batch_size.
+            #
+            # In the case when batch_size is not a multiple of mini_batch_size,
+            # we want to avoid having a tiny mini batch in the end. For example,
+            # when batch_size is 264 and mini_batch_size is 32, if
+            # mini_batch_size is not adjusted, there will be 8 mini batches of
+            # size 32 and 1 mini batch of size 8.
+            #
+            # In the above example, we will try to squeeze all the experience
+            # into 8 mini batches by adjusting the mini_batch_size to 33.
+            if batch_size % mini_batch_size > 0:
+                mini_batch_size = np.ceil(
+                    batch_size / (batch_size // mini_batch_size)).astype(int)
+
         def _make_time_major(nest):
             """Put the time dim to axis=0."""
             return alf.nest.map_structure(lambda x: x.transpose(0, 1), nest)
 
+        print(
+            f'batch_size = {batch_size} vs mini_batch_size = {mini_batch_size}'
+        )
         for u in range(num_updates):
+            # TODO(breakds): Currently the ``if`` below will also be executed
+            # for the cases when whole_replay_buffer_training = False and
+            # config.num_updates_per_train_iter > 0. This is not necessary and
+            # not desired because in that case the sampled trajectories are
```

Comment:
1. sampled trajectories are not guaranteed to be in random order (e.g. experiences from gather_all, even for it is not even guaranteed for get_batch())
2. even if it is, it is still better to randomize the order between updates